### PR TITLE
[Microservices] Fix redundant DB session creation for forwarded requests

### PR DESCRIPTION
**Pull request by moranbental**
FastAPI dependencies automatically create DB sessions when entering an endpoint. These sessions are redundant when the request is going to be forwarded to services. 
To address this, before entering the endpoint, we check if the request should be forwarded by verifying whether the service that received the request has the corresponding path and method in its `method_routes`.

Also, add a PATCH request to the method_routes to allow PATCH requests

https://iguazio.atlassian.net/browse/ML-8391

### CODE REVIEW

File: server/py/framework/utils/clients/messaging.py [hunk 1]
```diff
@@ -282,3 +289,20 @@ async def _resolve_request_kwargs_from_request(
 
         request_kwargs.update(**kwargs)
         return request_kwargs
+
+    @staticmethod
+    def _get_prefix_and_version(path: str):
+        match = PREFIX_GROUPING.match(path)
+        prefix = match.group(1)
+        version = match.group(2) or "v1"
+        return prefix, version
```
Comment (rokatyy) [inline server/py/framework/utils/clients/messaging.py]: if doesn't match, match will be None, as a result `match.group(1)` will raise an error. is it expected?

File: server/py/framework/utils/clients/messaging.py [hunk 2]
```diff
@@ -282,3 +289,20 @@ async def _resolve_request_kwargs_from_request(
...
+
+    def _prepare_request_data(self, method: str, path: str):
+        prefix, version = self._get_prefix_and_version(path)
+
+        # Remove the service and version prefix from the path
+        # The service prefix is to be replaced with the new service name
+        # The version will be re-added or default to v1 if not present
```
Comment (rokatyy) [inline server/py/framework/utils/clients/messaging.py]: This comment should be part of  `_get_prefix_and_version`

File: server/py/framework/utils/clients/messaging.py [hunk 3]
```diff
@@ -151,6 +145,19 @@ async def convert_requests_response_to_fastapi_response(
...
+    def should_forward_request(self, request: fastapi.Request) -> bool:
+        """
+        Checks whether the request should be forwarded to another service based on
+        the service and path being resolved.
+
+        :param request: The incoming FastAPI request.
+        :return: True if the request should be forwarded, False otherwise.
+        """
+        method = request.method
+        path = str(request.url.path)
```
Comment (rokatyy) [inline server/py/framework/utils/clients/messaging.py]: isn't it already a str?

**Review by rokatyy — COMMENTED** [review]
Looking very nice!

Couple of suggestions.

File: server/py/framework/utils/clients/messaging.py (hunk 2 above)
Comment (moranbental) [reply server/py/framework/utils/clients/messaging.py]: @rokatyy Actually, it is related to the `proxy_request` method, so I moved it there

**Review by liranbg — APPROVED** [review]

File: server/py/framework/utils/clients/messaging.py [hunk 4]
```diff
@@ -282,3 +293,19 @@ async def _resolve_request_kwargs_from_request(
...
         return request_kwargs
+
+    @staticmethod
+    def _get_prefix_and_version(path: str):
+        match = PREFIX_GROUPING.match(path)
+        if not match:
+            raise ValueError(f"Invalid path format: {path}")
+
+        prefix = match.group(1)
+        version = match.group(2) or "v1"  # default to v1 if not present
```
Comment (rokatyy) [inline server/py/framework/utils/clients/messaging.py]:
```suggestion
        # default to v1 if not present
        version = match.group(2) or "v1"
```

**Review by rokatyy — APPROVED** [review]
Minor nitpicking, otherwise approved 👍

File: server/py/framework/api/deps.py [hunk 5]
```diff
@@ -25,14 +25,19 @@
...
-    try:
-        db_session = framework.db.session.create_session()
-        yield db_session
-    finally:
-        if db_session:
-            framework.db.session.close_session(db_session)
+def get_db_session(
+    request: Request,
+) -> typing.Generator[typing.Optional[Session], None, None]:
+    if not request.app.extra.get("mlrun_service").should_forward_request(request):
```
Comment (TomerShor) [inline server/py/framework/api/deps.py]: Add a comment on WHY you are checking this.

File: server/py/framework/service/__init__.py [hunk 6]
```diff
@@ -349,6 +351,15 @@ async def _align_worker_state_with_chief_state(
             self._synchronize_with_chief_clusterization_spec.__name__
         )
 
+    def should_forward_request(self, request: fastapi.Request) -> bool:
```
Comment (TomerShor) [inline server/py/framework/service/__init__.py]:
```suggestion
    def is_forwarded_request(self, request: fastapi.Request) -> bool:
```

File: server/py/framework/utils/clients/messaging.py [hunk 7]
```diff
@@ -151,6 +149,19 @@ async def convert_requests_response_to_fastapi_response(
             media_type="application/json",
         )
 
+    def should_forward_request(self, request: fastapi.Request) -> bool:
```
Comment (TomerShor) [inline server/py/framework/utils/clients/messaging.py]:
```suggestion
    def is_forwarded_request(self, request: fastapi.Request) -> bool:
```

**Review by TomerShor — CHANGES_REQUESTED** [review]
Nice stuff.
Minor comments

**Pull request merged**